### DISCUSSION TITLE
Transaction in middleware mongodb

### DISCUSSION THREAD

**Discussion by conduongtong** [prisma/prisma#18844] [Q&A] [answered]
Can I use transaction with middleware?

For example my query:
```js
prisma.user.create({ data: { name: 'test' } });
```
Then inside middleware I will also create Profile:
```js
prisma.$use(async (params, next) => {
  const user = await next(params);
  const profile = await prisma.profile.create({ data: { user: user.id } });
  return user;
});
```

How can I rollback user  if profile not success?

**Comment by ludralph** [comment]
Hi @conduongtong 👋 

You can use the client extension as an alternative to middleware with a transaction in Prisma. Please do take a look at this [example](https://www.prisma.io/blog/client-extensions-preview-8t3w27xkrxxn#example-callback-free-interactive-transactions)

**Reply by conduongtong** [reply, marked as answer]
> Hi @conduongtong 👋
> 
> You can use the client extension as an alternative to middleware with a transaction in Prisma. Please do take a look at this [example](https://www.prisma.io/blog/client-extensions-preview-8t3w27xkrxxn#example-callback-free-interactive-transactions)

Sorry for late reply.
$extends not helping me because transaction will be call outside, and I need to use it in middleware for generic purpose and recursive.
Real example is create breadcrumbs. it have parent and children field, I want to transaction all of it, it will be use for many model like Page, Section,...
This is an example save as mine: https://github.com/prisma/prisma/discussions/17746

Currently this is my solution:
```js
prisma.$use(async (params, next) => {
  const user = await next(params);
  await prisma
    .$transaction(async ctx => {
      await ctx.profile.create({ data: { user: user.id } });
    })
    .catch(async error => {
      await prisma.user.delete({ data: { id: user.id } });
      throw error;
    });
  return user;
});
```
